src/lavinmq/queue/stream_queue_message_store.cr

```diff
@@ -88,6 +106,71 @@ module LavinMQ
         {msg_offset, segment, pos}
       end
 
+      def last_offset_by_consumer_tag(consumer_tag : String)
+        if pos = @consumer_offset_positions[consumer_tag]?
+          tx = @consumer_offsets.to_slice(pos, 8)
+          return IO::ByteFormat::SystemEndian.decode(Int64, tx)
+        end
+      end
+
+      private def restore_consumer_offset_positions : Hash(String, Int64)
+        positions = Hash(String, Int64).new
+        return positions if @consumer_offsets.size.zero?
+
+        loop do
+          ctag = AMQ::Protocol::ShortString.from_io(@consumer_offsets)
+          break if ctag.empty?
+          positions[ctag] = @consumer_offsets.pos
+          @consumer_offsets.skip(8)
+        rescue IO::EOFError
+          break
+        end
+        @consumer_offsets.pos = 0 if @consumer_offsets.pos == 1
+        @consumer_offsets.resize(@consumer_offsets.pos)
+        positions
+      end
+
+      def update_consumer_offset(consumer_tag : String, new_offset : Int64)
+        if pos = @consumer_offset_positions[consumer_tag]?
+          IO::ByteFormat::SystemEndian.encode(new_offset, @consumer_offsets.to_slice(pos, 8, false))
+        else
+          store_consumer_offset(consumer_tag, new_offset)
+        end
+      end
+
+      def store_consumer_offset(consumer_tag : String, new_offset : Int64)
+        @consumer_offsets.write_bytes AMQ::Protocol::ShortString.new(consumer_tag)
+        @consumer_offset_positions[consumer_tag] = @consumer_offsets.size
+        @consumer_offsets.write_bytes new_offset
+      end
+
+      def cleanup_consumer_offsets
+        return if @consumer_offsets.size.zero?
+
+        offsets_to_save = Hash(String, Int64).new
+        lowest_offset_in_stream, _seg, _pos = offset_at(@segments.first_key, 4u32)
+        @consumer_offset_positions.each do |ctag, _pos|
+          if offset = last_offset_by_consumer_tag(ctag)
+            offsets_to_save[ctag] = offset if offset >= lowest_offset_in_stream
+          end
+        end
+
+        @consumer_offset_positions = Hash(String, Int64).new
+        replace_offsets_file do
+          offsets_to_save.each do |ctag, offset|
+            store_consumer_offset(ctag, offset)
+          end
+        end
+      end
+
+      def replace_offsets_file(&)
+        old_consumer_offsets = @consumer_offsets
+        @consumer_offsets = MFile.new("#{@consumer_offset_path}.tmp", @consumer_offset_capacity)
+        yield # fill the new file with correct data in this block
+        File.rename "#{@consumer_offset_path}.tmp", @consumer_offset_path
```

Comment:
should implement `MFile#rename`, so that the internal `@path` is changed too. 
 
```suggestion
        mfile.rename @consumer_offset_path
```

Reply:
Implemented [here](https://github.com/cloudamqp/lavinmq/pull/661/commits/e70cc10091ee066d8daa1fce0d19ab0ddecfc590)